### ISSUE TITLE
Clang issues incorrect diagnostic when we have a catch blocks that catch both const char* and char*

### ISSUE THREAD

**Issue by shafik** [llvm/llvm-project#61177]
Given the following C++ code:

```cpp
#include <iostream>

int main() {
    const char arr[4] = "abc";

   try {
     throw arr;
   } catch (char *p) {
      std::cout << "char*\n";
   } catch (const char *p) {
      std::cout << "const char*\n";
   }
}
```

clang issues the following diagnostic: https://godbolt.org/z/j38xbs5oE

```console
<source>:10:19: warning: exception of type 'const char *' will be caught by earlier handler [-Wexceptions]
   } catch (const char *p) {
                  ^
<source>:8:13: note: for type 'char *'
   } catch (char *p) {
            ^
1 warning generated.
ASM generation compiler returned: 0
<source>:10:19: warning: exception of type 'const char *' will be caught by earlier handler [-Wexceptions]
   } catch (const char *p) {
                  ^
<source>:8:13: note: for type 'char *'
   } catch (char *p) {
            ^
1 warning generated.
Execution build compiler returned: 0
Program returned: 0
const char*
```

We we can see from the full output is incorrect. It is likely because the wording in [expr.throw p2](https://eel.is/c++draft/expr.throw#2) has a defect which says that we throw `arr` the type should be `char *` but this it not intended and it subject of [Defect Report 2699](https://wg21.link/cwg2699) which is not published yet.

**Comment by shafik** [comment]
CC @AaronBallman
